Integrate ARMv8 PMULL+EOR3 crc32 algorithm from fast-crc32

nmoinvaz opened the pull request:
From discussion #1998. 

I have integrated Peter Cawley's @corsix fast-crc32 algorithms. These speed up CRC32 operations on Apple ARMv8 cpus. It allows the CPU to parallelize the operations, where as the normal `_crc32*` intrinsics aren't parallelizable because each call depends on the output of a previous operation.

I asked AI to improve the comments in both algorithms. Benchmarks below.

nmoinvaz commented:
## Benchmarks
```
Running /Users/nathan/Source/zlib-ng/build/test/benchmarks/benchmark_zlib
Run on (8 X 24 MHz CPU s)
CPU Caches:
  L1 Data 64 KiB
  L1 Instruction 128 KiB
  L2 Unified 4096 KiB (x8)
Load Average: 3.17, 2.52, 2.38
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
crc32/braid/1                        2.71 ns         2.71 ns    244666275
crc32/braid/8                        17.5 ns         17.4 ns     40065478
crc32/braid/12                       25.7 ns         25.7 ns     27174757
crc32/braid/16                       34.5 ns         34.5 ns     20462033
crc32/braid/32                       67.9 ns         67.9 ns     10326464
crc32/braid/64                        127 ns          126 ns      5593869
crc32/braid/512                       198 ns          195 ns      3308895
crc32/braid/4096                      618 ns          618 ns      1154392
crc32/braid/32768                    4233 ns         4233 ns       167311
crc32/braid/262144                  34262 ns        34252 ns        20599
crc32/braid/4194304                537680 ns       537485 ns         1308
crc32/chorba_c/1                     2.74 ns         2.73 ns    259969843
crc32/chorba_c/8                     17.4 ns         17.4 ns     40252556
crc32/chorba_c/12                    25.7 ns         25.7 ns     27250714
crc32/chorba_c/16                    34.2 ns         34.2 ns     20315234
crc32/chorba_c/32                    67.9 ns         67.9 ns     10331341
crc32/chorba_c/64                     125 ns          125 ns      5584230
crc32/chorba_c/512                    179 ns          179 ns      3903679
crc32/chorba_c/4096                   571 ns          571 ns      1224997
crc32/chorba_c/32768                 2444 ns         2444 ns       288412
crc32/chorba_c/262144               28783 ns        28777 ns        24372
crc32/chorba_c/4194304             202030 ns       202030 ns         3464
crc32/armv8/1                        1.49 ns         1.49 ns    470221542
crc32/armv8/8                        1.74 ns         1.74 ns    403053998
crc32/armv8/12                       2.24 ns         2.24 ns    314061897
crc32/armv8/16                       2.00 ns         2.00 ns    352975584
crc32/armv8/32                       3.48 ns         3.48 ns    201957837
crc32/armv8/64                       6.45 ns         6.45 ns    108109778
crc32/armv8/512                      48.2 ns         48.2 ns     14543128
crc32/armv8/4096                      382 ns          382 ns      1832379
crc32/armv8/32768                    3052 ns         3052 ns       229786
crc32/armv8/262144                  24393 ns        24388 ns        28678
crc32/armv8/4194304                390667 ns       390577 ns         1797
crc32/armv8_pmull/1                  2.50 ns         2.50 ns    281108693
crc32/armv8_pmull/8                  2.52 ns         2.52 ns    280522895
crc32/armv8_pmull/12                 4.23 ns         4.23 ns    166063697
crc32/armv8_pmull/16                 2.76 ns         2.76 ns    253372570
crc32/armv8_pmull/32                 6.77 ns         6.76 ns    103232657
crc32/armv8_pmull/64                 8.43 ns         8.43 ns     82977715
crc32/armv8_pmull/512                32.2 ns         32.2 ns     21107352
crc32/armv8_pmull/4096                130 ns          130 ns      5446242
crc32/armv8_pmull/32768               694 ns          694 ns      1021406
crc32/armv8_pmull/262144             5051 ns         5050 ns       139151
crc32/armv8_pmull/4194304           79166 ns        79157 ns         8858
crc32/armv8_pmull_eor3/1             2.24 ns         2.24 ns    314202867
crc32/armv8_pmull_eor3/8             2.48 ns         2.48 ns    282345728
crc32/armv8_pmull_eor3/12            4.22 ns         4.22 ns    166174079
crc32/armv8_pmull_eor3/16            3.23 ns         3.23 ns    217309645
crc32/armv8_pmull_eor3/32            6.59 ns         6.59 ns    107516972
crc32/armv8_pmull_eor3/64            8.17 ns         8.17 ns     86437894
crc32/armv8_pmull_eor3/512           30.2 ns         30.2 ns     23273288
crc32/armv8_pmull_eor3/4096          96.1 ns         96.0 ns      7322942
crc32/armv8_pmull_eor3/32768          391 ns          391 ns      1795581
crc32/armv8_pmull_eor3/262144        2652 ns         2652 ns       264047
crc32/armv8_pmull_eor3/4194304      41019 ns        41003 ns        17095
```

nmoinvaz commented on CMakeLists.txt:
```diff
@@ -825,6 +826,29 @@ if(WITH_OPTIM)
+            endif()
+            if(NOT HAVE_ARMV8_PMULL_INTRIN AND NOT HAVE_ARMV8_PMULL_EO3_INTRIN)
+                set(WITH_ARMV8_PMULL OFF)
+            endif()
```
Fixed.

nmoinvaz commented on arch/arm/crc32_armv8_pmull.c:
```diff
@@ -0,0 +1,164 @@
+    return r;
+}
+
+Z_INTERNAL Z_TARGET_PMULL uint32_t crc32_armv8_pmull(uint32_t crc, const char* buf, size_t len) {
```
Fixed.

nmoinvaz commented on arch/arm/arm_functions.h:
```diff
@@ -66,6 +72,19 @@ void slide_hash_armv6(deflate_state *s);
+#    define native_crc32_fold_final crc32_fold_pmull_final
+#    undef native_crc32_fold_reset
+#    define native_crc32_fold_reset crc32_fold_pmull_reset
+#  endif
```
Fixed.

nmoinvaz commented on arch/arm/arm_functions.h:
```diff
@@ -24,6 +24,12 @@ uint32_t crc32_armv8(uint32_t crc, const uint8_t *buf, size_t len);
+#endif
+#ifdef ARM_PMULL_EOR3
+uint32_t crc32_armv8_pmull_eor3(uint32_t crc, const uint8_t *buf, size_t len);
+#endif
```
Fixed.

nmoinvaz commented on arch/arm/crc32_armv8_pmull_eor3.c:
```diff
@@ -0,0 +1,196 @@
+    return r;
+}
+
+Z_INTERNAL Z_TARGET_PMULL_EOR3 uint32_t crc32_armv8_pmull_eor3(uint32_t crc, const char *buf, uint32_t len) {
```
Fixed.

nmoinvaz commented on arch/arm/crc32_armv8_pmull.c:
```diff
@@ -0,0 +1,164 @@
+        size_t blk = len / 112;               /* Number of 112-byte blocks */
+        size_t klen = blk * 16;               /* Scalar stride per CRC lane */
+        const char* buf2 = buf + klen * 4;    /* Vector data starts after scalar lanes */
+        const char* limit = buf + klen - 32;
```
Fixed.

nmoinvaz commented on test/benchmarks/benchmark_crc32.cc:
```diff
@@ -76,6 +76,12 @@ BENCHMARK_CRC32(chorba_c, crc32_chorba, 1);
+#endif
+#ifdef ARM_PMULL_EOR3
+BENCHMARK_CRC32(armv8_pmull_eor3, crc32_armv8_pmull_eor3, test_cpu_features.arm.has_pmull && test_cpu_features.arm.has_eor3);
+#endif
```
Fixed.

nmoinvaz commented on functable.c:
```diff
@@ -252,7 +252,18 @@ static int init_functable(void) {
+    if (cf.arm.has_pmull && cf.arm.has_eor3) {
+        ft.crc32 = &crc32_armv8_pmull_eor3;
+    }
+#endif
```
Fixed.

ccawley2011 commented on cmake/detect-intrinsics.cmake:
```diff
@@ -56,6 +56,63 @@ macro(check_armv8_compiler_flag)
+    # Check whether compiler supports ARMv8 PMULL intrinsics
+    set(CMAKE_REQUIRED_FLAGS "${PMULLFLAG} ${NATIVEFLAG} ${ZNOLTOFLAG}")
+    check_c_source_compiles(
+        "#include <arm_neon.h>
```
```suggestion
        "#if defined(_MSC_VER) && (defined(_M_ARM64) || defined(_M_ARM64EC))
        #  include <arm64_neon.h>
        #else
        #  include <arm_neon.h>
        #endif
```

nmoinvaz replied:
Fixed.

ccawley2011 commented on cmake/detect-intrinsics.cmake:
```diff
@@ -56,6 +56,63 @@ macro(check_armv8_compiler_flag)
+    # Check whether compiler supports ARMv8 PMULL + EOR3 intrinsics
+    set(CMAKE_REQUIRED_FLAGS "${PMULLEOR3FLAG} ${NATIVEFLAG} ${ZNOLTOFLAG}")
+    check_c_source_compiles(
+        "#include <arm_neon.h>
```
```suggestion
        "#if defined(_MSC_VER) && (defined(_M_ARM64) || defined(_M_ARM64EC))
        #  include <arm64_neon.h>
        #else
        #  include <arm_neon.h>
        #endif
```

nmoinvaz replied:
Fixed.

ccawley2011 commented on cmake/detect-intrinsics.cmake:
```diff
@@ -56,6 +56,63 @@ macro(check_armv8_compiler_flag)
+            return vmull_p64(a, b);
+        }
+        uint64x2_t g(uint64x2_t a, uint64x2_t b, uint64x2_t c) {
+            return veor3q_u64(a, b, c);
```
This checks for the `vmull_p64` and `veor3q_u64` intrinsics, but the code doesn't appear to be using them.

nmoinvaz replied:
`veor3q_u64` is used in crc32_armv8_pmull_eor3.c.

nmoinvaz replied:
I have updated the code to now use `vmull_p64` intrinsic.

ccawley2011 commented on CMakeLists.txt:
```diff
@@ -119,7 +119,7 @@ set_property(CACHE WITH_SANITIZER PROPERTY STRINGS "Memory" "Address" "Undefined
 if(BASEARCH_ARM_FOUND)
     cmake_dependent_option(WITH_ARMV6 "Build with ARMv6 SIMD" ON "NOT ARCH MATCHES \"aarch64\"" OFF)
-    option(WITH_ARMV8 "Build with ARMv8 CRC32 intrinsics" ON)
+    option(WITH_ARMV8 "Build with ARMv8 intrinsics" ON)
```
It might be worth having a separate option for building with PMULL support.

nmoinvaz replied:
I thought about that, but then I thought I might have to make another one for PMULL+EOR3, and it just seemed like it was getting more complicated. I figure if somebody needs that level of granularity we can always add it later.

nmoinvaz commented on configure:
```diff
@@ -2090,6 +2152,44 @@ EOF
+
+                    if test $HAVE_ARMV8_INTRIN -eq 1; then
+                        CFLAGS="${CFLAGS} -DARM_CRC32_INTRIN"
+                        SFLAGS="${SFLAGS} -DARM_CRC32_INTRIN"
+                    fi
+
+                    ARCH_STATIC_OBJS="${ARCH_STATIC_OBJS} crc32_armv8.o"
+                    ARCH_SHARED_OBJS="${ARCH_SHARED_OBJS} crc32_armv8.lo"
+
+                    # Check for PMULL support (requires NEON too)
+                    if test $NEON_AVAILABLE -eq 1; then
+                        check_armv8_pmull_intrinsics
+
+                        if test $HAVE_PMULL_INTRIN -eq 1; then
+                            CFLAGS="${CFLAGS} -DARM_PMULL"
+                            SFLAGS="${SFLAGS} -DARM_PMULL"
+                            ARCH_STATIC_OBJS="${ARCH_STATIC_OBJS} crc32_armv8_pmull.o"
+                            ARCH_SHARED_OBJS="${ARCH_SHARED_OBJS} crc32_armv8_pmull.lo"
+
+                            # Check for PMULL+EOR3 support (ARMv8.2-A SHA3 extension)
+                            check_armv8_pmull_eor3_intrinsics
+
+                            if test $HAVE_PMULL_EOR3_INTRIN -eq 1; then
+                                CFLAGS="${CFLAGS} -DARM_PMULL_EOR3"
+                                SFLAGS="${SFLAGS} -DARM_PMULL_EOR3"
+                                ARCH_STATIC_OBJS="${ARCH_STATIC_OBJS} crc32_armv8_pmull_eor3.o"
+                                ARCH_SHARED_OBJS="${ARCH_SHARED_OBJS} crc32_armv8_pmull_eor3.lo"
+                            fi
+                        fi
+                    fi
+                fi
+            fi
```
Fixed.

nmoinvaz commented:
~~I will rebase after PR #2028 lands.~~ Not necessary.

nmoinvaz commented on arch/arm/crc32_armv8_pmull.c:
```diff
@@ -0,0 +1,158 @@
+            crc3 = __crc32d(crc3, *(const uint64_t*)(buf + klen * 3 + 8));
+            buf += 16;
+            buf2 += 48;
+        }
```
should be fixed now.

nmoinvaz commented on arch/arm/crc32_armv8_pmull_eor3.c:
```diff
@@ -0,0 +1,182 @@
+            crc2 = __crc32d(crc2, *(const uint64_t*)(buf + klen * 2 + 8));
+            buf += 16;
+            buf2 += 144;
+        }
```
should be fixed now.

nmoinvaz commented on arch/arm/arm_features.c:
```diff
@@ -59,6 +59,111 @@ static int arm_has_crc32(void) {
+    return haspmull;
+#elif defined(__APPLE__)
+    /* macOS doesn't expose crypto/pmull directly, check for neon_hpfp as proxy */
+    /* Or just return 1 for AArch64 as all Apple Silicon has crypto extensions */
+#  if defined(__aarch64__) || defined(_M_ARM64)
+    return 1; /* All Apple AArch64 CPUs have crypto extensions */
+#  else
+    int haspmull;
+    size_t size = sizeof(haspmull);
+    return sysctlbyname("hw.optional.neon_hpfp", &haspmull, &size, NULL, 0) == 0
+      && haspmull == 1;
+#  endif
```
should be fixed now.

nmoinvaz commented on arch/arm/crc32_armv8_pmull.h:
```diff
@@ -0,0 +1,52 @@
+
+#include <arm_acle.h>
+#include <arm_neon.h>
+
```
should be fixed now.

nmoinvaz commented:
Rebased.

Dead2 commented:
Tested this on RPI5:
```
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
crc32/armv8/1                        2.17 ns         2.17 ns   1287449066
crc32/armv8/8                        4.35 ns         4.35 ns    643726859
crc32/armv8/12                       4.35 ns         4.35 ns    643742153
crc32/armv8/16                       4.78 ns         4.78 ns    585211951
crc32/armv8/32                       6.52 ns         6.52 ns    429155011
crc32/armv8/64                       8.26 ns         8.26 ns    338807708
crc32/armv8/512                      36.3 ns         36.3 ns     76949440
crc32/armv8/4096                      236 ns          236 ns     11854793
crc32/armv8/32768                    1795 ns         1795 ns      1559626
crc32/armv8/262144                  15298 ns        15298 ns       182977
crc32/armv8/4194304                250046 ns       250037 ns        11169
crc32/armv8_pmull/1                  4.78 ns         4.78 ns    585206246
crc32/armv8_pmull/8                  4.78 ns         4.78 ns    585221125
crc32/armv8_pmull/12                 7.39 ns         7.39 ns    378672344
crc32/armv8_pmull/16                 5.22 ns         5.22 ns    536455506
crc32/armv8_pmull/32                 10.9 ns         10.9 ns    257497763
crc32/armv8_pmull/64                 14.2 ns         14.2 ns    196632564
crc32/armv8_pmull/512                93.9 ns         93.9 ns     29808363
crc32/armv8_pmull/4096                267 ns          267 ns     10461153
crc32/armv8_pmull/32768              1329 ns         1329 ns      2107212
crc32/armv8_pmull/262144             9545 ns         9545 ns       293234
crc32/armv8_pmull/4194304          343378 ns       343364 ns         8079
crc32/armv8_pmull_eor3/1       ERROR OCCURRED: 'CPU does not support armv8_pmull_eor3'
*snip*
```

Unfortunately it seems to be slower than the normal armv8 one for this cpu.
Deflatebench is slightly slower, but not really within confidence levels, although the above tells me it is probably true.

Maybe we need improved detection for whether or not this is a benefit for the specific cpu. Or only enable it on Apple cpus for now, until someone finds others that benefit.

Edit: Interestingly it is faster for 32k and 256k, just not any of the others. 512 is nearly 3x as slow for example.

nmoinvaz commented:
From what I'm seeing, pmull method will not perform well on Cortex-A series because it only has 1 pmull lane shared between multiple execution units.

nmoinvaz commented:
All Apple ARM CPU's have multiple PMULL lanes, so the easiest thing to do would be to enable this only for macOS. OR we could have a function that reads the cpu id and makes a determination. Most of the armv8 that also support SHA3 have multiple PMULL lanes, however armv9 can have a single PMULL lane and also have SHA3 support.

nmoinvaz commented:
https://gist.github.com/nmoinvaz/b56489b6643156df798ea8f04d1ceefd
@dead2 what do you think?

nmoinvaz committed:
Check CPU info for fast PMULL support.

armv8_pmull_eor3 is beneficial only if the CPU has multiple PMULL
execution units.

Co-authored-by: Adam Stylinski <kungfujesus06@gmail.com>

Dead2 commented:
> [gist.github.com/nmoinvaz/b56489b6643156df798ea8f04d1ceefd](https://gist.github.com/nmoinvaz/b56489b6643156df798ea8f04d1ceefd) @Dead2 what do you think?

Is there any other way to detect arm cpu model/manufacturer?

/proc/cpuinfo might not be accessible as a non-root user.
It also requires a pass through every active LSM, even if access ends up permitted, so it's not of the fastest. It would also flag any failed attempt at accessing /proc and-or /proc/cpuinfo in the audit log, and attribute it to the application directly/indirectly linking in zlib-ng.

Looks like it might be possible for Aarch64, seems like the MIDR_EL1 register can give an id of who manufactured the cpu.
```
# The core type (via MIDR_EL1)
MIDR_EL1 tells you:
- Implementer
- PartNum (core type)
- Variant + revision
Example: Cortex-A55 vs Cortex-A76 vs Neoverse V1, etc.
```
A quick google search did not reveal how to do this though.

KungFuJesus commented:
I agree identifying if apple silicon by a cpuid-like instruction is the way to go. Parsing /proc/cpuinfo is not portable outside of Linux, though. We'd have to figure out how to extract a vendor string or id similar to what we do with x86.

KungFuJesus commented:
https://elixir.bootlin.com/linux/v6.13/source/arch/arm64/kernel/cpuinfo.c#L277

KungFuJesus commented:
Tested it in the M2 - it's about an order of magnitude faster for the largest buffer size.  It seems that Apple is the only Arm implementation that actually implements clmul for speed.  Funny, given that this was meant to make cryptographic operations faster.

```
--------------------------------------------------------------
Benchmark                    Time             CPU   Iterations
--------------------------------------------------------------
crc32/armv8/4194304     524795 ns       523766 ns         1330
```

vs
```
crc32/armv8_pmull_eor3/4194304      53519 ns        53443 ns        13099
```

KungFuJesus commented:
@nmoinvaz try this 
[0001-Add-methods-to-detect-if-on-apple-silicon.patch](https://github.com/user-attachments/files/24040072/0001-Add-methods-to-detect-if-on-apple-silicon.patch)

nmoinvaz commented:
MIDR_EL1 is not allowed from user mode at least on M3 it gives seg fault.

nmoinvaz commented:
@KungFuJesus Did you see my last commit?

KungFuJesus commented:
I did but only after I wrote that.  Asahi seems able to do MIDR_EL1 in user mode, what are you testing on?  You'll get a sigill if the CPUID HWCAP is not supported but, at least on Mac OS, we can safely assume you're on apple silicon if you're running an aarch64 binary from mac os.

KungFuJesus commented:
Hah, oh, you have a somewhat working whitelist there.  The MIDR_EL1 system register has not produced issues for me, but if you're testing on mac OS or windows I guess it's possible they hide access to that?

nmoinvaz commented:
If we only enable this on macOS it can be a compile time ifdef.

KungFuJesus commented:
> If we only enable this on macOS it can be a compile time ifdef.

Check out my patch.  I unconditionally enable it compile time if compiling on Mac OS and guard it otherwise behind that check.  We want this to work on Apple Silicon on Linux, too.  I think this gets us the best of both worlds.

nmoinvaz commented:
<img width="446" height="195" alt="image" src="https://github.com/user-attachments/assets/7533785f-3435-4457-8b65-b6562e2f7bf0" />

This is Apple M3, macOS 15.6.1.

KungFuJesus commented:
Are you trying that on Linux? It works perfectly for me on asahi. My patch only checks if it's apple silicon on Linux and assumes it is if on Mac OS.

nmoinvaz commented on arch/arm/arm_features.c:
```diff
@@ -121,6 +224,96 @@ static inline int arm_has_simd(void) {
+#  if defined(__linux__)
+    /* We have to support the CPUID feature in HWCAP */
+    if (!(getauxval(AT_HWCAP) & HWCAP_CPUID))
+        return 0;
+#  elif defined(__FreeBSD__) || defined(__OpenBSD__)
+    unsigned long hwcap = 0;
+    elf_aux_info(AT_HWCAP, &hwcap, sizeof(hwcap));
+    if (!(hwcap & HWCAP_CPUID))
+        return 0;
+#  endif
+    uint64_t midr = read_midr_el1();
+    uint32_t implementer = MIDR_IMPLEMENTOR(midr);
+    uint32_t part = MIDR_PARTNUM(midr);
+
+    if (implemented == ARM_IMPLEMENTER_APPLE) {
+        /* All Apple Silicon (M1+) have fast PMULL */
+        return 1;
+    } else if (implementer == ARM_IMPLEMENTER_ARM) {
+        /* ARM Cortex-X and Neoverse V/N2 series have multi-lane PMULL */
+        switch (part) {
+            case ARM_PART_CORTEX_X1:
+            case ARM_PART_CORTEX_X1C:
+            case ARM_PART_CORTEX_X2:
+            case ARM_PART_CORTEX_X3:
+            case ARM_PART_CORTEX_X4:
+            case ARM_PART_CORTEX_X925:
+            case ARM_PART_NEOVERSE_N2:
+            case ARM_PART_NEOVERSE_V1:
+            case ARM_PART_NEOVERSE_V2:
+            case ARM_PART_NEOVERSE_V3:
+                return 1;
+        }
+    } else if (implementer == ARM_IMPLEMENTER_QUALCOMM) {
+        /* Qualcomm Oryon (Snapdragon X Elite/Plus) has fast PMULL */
+        if (part == QUALCOMM_PART_ORYON)
+            return 1;
+    }
+#endif
+    return 0;
+}
```
Fixed

KungFuJesus reviewed (approved):
LGTM

nmoinvaz commented:
I am wondering if I should remove the `crc32_armv8_pmull` implementation since I think it finding a aarch64 CPU with fast PMULL (multiple PMULL execution units) without EOR3 (crypto extensions) would be rare.

KungFuJesus commented:
If one exists it's probably worth keeping, even if it requires high end arm servers

nmoinvaz commented:
I asked AI (I know sometimes unreliable) and it couldn't find any. But if Dead2's RPI5 with Arm Cortex A76 (missing crypto extensions) doesn't perform well, I am not sure it is worth keeping around.

Dead2 commented:
What I'm left wondering is whether it is possible to optimize for the single-lane PMULL? Interleaving PMULL and regular scalar code for example. I have not more than glanced at this code so far, so it might be impossible for all I know.

nmoinvaz commented:
@Dead2 can you try this function on your RP, it passes gtest_lib. https://gist.github.com/nmoinvaz/889dafb1f9c182b59192ec3d45729a55

Dead2 commented:
> @Dead2 can you try this function on your RP, it passes gtest_lib. [gist.github.com/nmoinvaz/889dafb1f9c182b59192ec3d45729a55](https://gist.github.com/nmoinvaz/889dafb1f9c182b59192ec3d45729a55)

Sorry, that is even worse for some reason..
```
crc32/armv8/1                        2.17 ns         2.17 ns   1287459366
crc32/armv8/8                        4.35 ns         4.35 ns    643746622
crc32/armv8/12                       4.35 ns         4.35 ns    643721049
crc32/armv8/16                       4.78 ns         4.78 ns    585215259
crc32/armv8/32                       6.52 ns         6.52 ns    429162453
crc32/armv8/64                       8.26 ns         8.26 ns    338806554
crc32/armv8/512                      36.4 ns         36.4 ns     77255501
crc32/armv8/4096                      236 ns          236 ns     11854741
crc32/armv8/32768                    1795 ns         1795 ns      1559688
crc32/armv8/262144                  15301 ns        15300 ns       182995
crc32/armv8/4194304                250823 ns       250814 ns        11123

crc32/armv8_pmull/1                  4.78 ns         4.78 ns    585219880
crc32/armv8_pmull/8                  4.35 ns         4.35 ns    643733054
crc32/armv8_pmull/12                 6.96 ns         6.96 ns    402340142
crc32/armv8_pmull/16                 6.52 ns         6.52 ns    429154597
crc32/armv8_pmull/32                 10.4 ns         10.4 ns    268226216
crc32/armv8_pmull/64                 13.9 ns         13.9 ns    201162761
crc32/armv8_pmull/512                62.6 ns         62.6 ns     44703642
crc32/armv8_pmull/4096                452 ns          452 ns      6189750
crc32/armv8_pmull/32768              3570 ns         3570 ns       784257
crc32/armv8_pmull/262144            28516 ns        28515 ns        98187
crc32/armv8_pmull/4194304          462647 ns       462633 ns         6047
```

nmoinvaz commented:
Ah well, it was worth a shot. I'm not an expert on this stuff. At least now the PR will detect the proper CPU for PMULL.

nmoinvaz commented:
I can remove the pmull (without eor3) implementation in a separate commit, so we have the original..

nmoinvaz commented on arch/arm/crc32_armv8_pmull_eor3.c:
```diff
@@ -0,0 +1,248 @@
+#if defined(ARM_PMULL_EOR3)
+#include "zbuild.h"
+#include "zutil.h"
+#include "acle_intrins.h"
+#include "crc32.h"
+
+#include "acle_intrins.h"
+#include "neon_intrins.h"
+
```
Fixed.

nmoinvaz commented on arch/arm/arm_features.c:
```diff
@@ -71,6 +71,109 @@ static int arm_has_crc32(void) {
+    return 0;
+#    endif
+#  endif
+#elif defined(__FreeBSD__) && defined(__aarch64__)
+    /* Check for AES feature as PMULL is part of crypto extension */
+    return getenv("QEMU_EMULATING") == NULL
+      && ID_AA64ISAR0_AES_VAL(READ_SPECIALREG(id_aa64isar0_el1)) >= ID_AA64ISAR0_AES_BASE;
+#elif defined(__OpenBSD__) && defined(__aarch64__)
+    int haspmull = 0;
+    int isar0_mib[] = { CTL_MACHDEP, CPU_ID_AA64ISAR0 };
+    uint64_t isar0 = 0;
+    size_t len = sizeof(isar0);
+    if (sysctl(isar0_mib, 2, &isar0, &len, NULL, 0) != -1) {
+      if (ID_AA64ISAR0_AES(isar0) >= ID_AA64ISAR0_AES_BASE)
+          haspmull = 1;
+    }
+    return haspmull;
+#elif defined(__APPLE__)
+#  if defined(__aarch64__) || defined(_M_ARM64)
+    int haspmull;
+    size_t size = sizeof(haspmull);
+    return sysctlbyname("hw.optional.arm.FEAT_PMULL", &haspmull, &size, NULL, 0) == 0
+      && haspmull == 1;
+#  else
+    return 0;
+#  endif
+#elif defined(_WIN32)
+    /* Windows checks for crypto/AES support */
+#  ifdef PF_ARM_V8_CRYPTO_INSTRUCTIONS_AVAILABLE
+    return IsProcessorFeaturePresent(PF_ARM_V8_CRYPTO_INSTRUCTIONS_AVAILABLE);
+#  else
+    return 0;
+#  endif
+#elif defined(__ARM_FEATURE_CRYPTO) || defined(__ARM_FEATURE_AES)
+    /* Compile-time check */
+    return 1;
+#else
+    return 0;
+#endif
+}
```
Fixed.

nmoinvaz commented on arch/arm/arm_features.c:
```diff
@@ -121,6 +224,85 @@ static inline int arm_has_simd(void) {
+    if (!(getauxval(AT_HWCAP) & HWCAP_CPUID))
+        return 0;
+#  elif defined(__FreeBSD__) || defined(__OpenBSD__)
+    unsigned long hwcap = 0;
+    elf_aux_info(AT_HWCAP, &hwcap, sizeof(hwcap));
+    if (!(hwcap & HWCAP_CPUID))
+        return 0;
+#  endif
+    uint64_t midr;
+    __asm__ ("mrs %0, midr_el1" : "=r" (midr));
+
+    uint32_t implementer = MIDR_IMPLEMENTOR(midr);
+    uint32_t part = MIDR_PARTNUM(midr);
+
+    if (implementer == ARM_IMPLEMENTER_APPLE) {
+        /* All Apple Silicon (M1+) have fast PMULL */
+        return 1;
+    } else if (implementer == ARM_IMPLEMENTER_ARM) {
+        /* ARM Cortex-X and Neoverse V/N2 series have multi-lane PMULL */
+        switch (part) {
+            case ARM_PART_CORTEX_X1:
+            case ARM_PART_CORTEX_X1C:
+            case ARM_PART_CORTEX_X2:
+            case ARM_PART_CORTEX_X3:
+            case ARM_PART_CORTEX_X4:
+            case ARM_PART_CORTEX_X925:
+            case ARM_PART_NEOVERSE_N2:
+            case ARM_PART_NEOVERSE_V1:
+            case ARM_PART_NEOVERSE_V2:
+            case ARM_PART_NEOVERSE_V3:
+                return 1;
+        }
+    } else if (implementer == ARM_IMPLEMENTER_QUALCOMM) {
+        /* Qualcomm Oryon (Snapdragon X Elite/Plus) has fast PMULL */
+        if (part == QUALCOMM_PART_ORYON)
+            return 1;
+    }
+#endif
+    return 0;
+}
```
Fixed.

nmoinvaz commented:
Here are benchmarks again on Apple M3 macOS 15.6.1
```
Running /Users/nathan/Source/zlib-ng/build/test/benchmarks/benchmark_zlib
Run on (8 X 24 MHz CPU s)
CPU Caches:
  L1 Data 64 KiB
  L1 Instruction 128 KiB
  L2 Unified 4096 KiB (x8)
Load Average: 3.05, 2.19, 1.87
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
crc32/braid/1                        2.75 ns         2.75 ns    255001676
crc32/braid/8                        17.4 ns         17.4 ns     40444660
crc32/braid/12                       25.9 ns         25.9 ns     27186366
crc32/braid/16                       34.2 ns         34.2 ns     20452646
crc32/braid/32                       68.0 ns         68.0 ns     10306243
crc32/braid/64                        125 ns          125 ns      5574225
crc32/braid/512                       190 ns          190 ns      3673287
crc32/braid/4096                      624 ns          624 ns      1132063
crc32/braid/32768                    4242 ns         4242 ns       165575
crc32/braid/262144                  34588 ns        34588 ns        20335
crc32/braid/4194304                555650 ns       550634 ns         1292
crc32/chorba_c/1                     2.71 ns         2.71 ns    258556369
crc32/chorba_c/8                     17.6 ns         17.6 ns     39688389
crc32/chorba_c/12                    26.0 ns         26.0 ns     26477141
crc32/chorba_c/16                    38.1 ns         35.5 ns     20279745
crc32/chorba_c/32                    68.2 ns         68.2 ns     10192047
crc32/chorba_c/64                     126 ns          125 ns      5571385
crc32/chorba_c/512                    196 ns          194 ns      3594592
crc32/chorba_c/4096                   602 ns          602 ns      1157943
crc32/chorba_c/32768                 2570 ns         2512 ns       284863
crc32/chorba_c/262144               32374 ns        31273 ns        22562
crc32/chorba_c/4194304             206799 ns       206619 ns         3403
crc32/armv8/1                        1.60 ns         1.60 ns    440005280
crc32/armv8/8                        1.86 ns         1.86 ns    379698086
crc32/armv8/12                       2.41 ns         2.41 ns    296019385
crc32/armv8/16                       2.26 ns         2.20 ns    331802303
crc32/armv8/32                       3.72 ns         3.72 ns    186580094
crc32/armv8/64                       6.83 ns         6.83 ns    105271073
crc32/armv8/512                      48.1 ns         48.1 ns     14498127
crc32/armv8/4096                      428 ns          413 ns      1801551
crc32/armv8/32768                    3104 ns         3104 ns       219020
crc32/armv8/262144                  25211 ns        25204 ns        28145
crc32/armv8/4194304                408531 ns       407430 ns         1715
crc32/armv8_pmull_eor3/1             1.87 ns         1.87 ns    364332854
crc32/armv8_pmull_eor3/8             2.20 ns         2.16 ns    327191482
crc32/armv8_pmull_eor3/12            2.66 ns         2.66 ns    266776426
crc32/armv8_pmull_eor3/16            2.39 ns         2.39 ns    295468360
crc32/armv8_pmull_eor3/32            3.72 ns         3.72 ns    189334516
crc32/armv8_pmull_eor3/64            6.73 ns         6.73 ns    106496273
crc32/armv8_pmull_eor3/512           32.4 ns         32.4 ns     21709196
crc32/armv8_pmull_eor3/4096          99.3 ns         99.3 ns      7055598
crc32/armv8_pmull_eor3/32768          411 ns          405 ns      1750525
crc32/armv8_pmull_eor3/262144        2741 ns         2703 ns       249862
crc32/armv8_pmull_eor3/4194304      41086 ns        41068 ns        16103
```

nmoinvaz commented on arch/arm/arm_features.c:
```diff
@@ -121,6 +224,85 @@ static inline int arm_has_simd(void) {
+    if (!(getauxval(AT_HWCAP) & HWCAP_CPUID))
+        return 0;
+#  elif defined(__FreeBSD__) || defined(__OpenBSD__)
+    unsigned long hwcap = 0;
+    elf_aux_info(AT_HWCAP, &hwcap, sizeof(hwcap));
+    if (!(hwcap & HWCAP_CPUID))
+        return 0;
+#  endif
+    uint64_t midr;
+    __asm__ ("mrs %0, midr_el1" : "=r" (midr));
+
+    uint32_t implementer = MIDR_IMPLEMENTOR(midr);
+    uint32_t part = MIDR_PARTNUM(midr);
+
+    if (implementer == ARM_IMPLEMENTER_APPLE) {
+        /* All Apple Silicon (M1+) have fast PMULL */
+        return 1;
+    } else if (implementer == ARM_IMPLEMENTER_ARM) {
+        /* ARM Cortex-X and Neoverse V/N2 series have multi-lane PMULL */
+        switch (part) {
+            case ARM_PART_CORTEX_X1:
+            case ARM_PART_CORTEX_X1C:
+            case ARM_PART_CORTEX_X2:
+            case ARM_PART_CORTEX_X3:
+            case ARM_PART_CORTEX_X4:
+            case ARM_PART_CORTEX_X925:
+            case ARM_PART_NEOVERSE_N2:
+            case ARM_PART_NEOVERSE_V1:
+            case ARM_PART_NEOVERSE_V2:
+            case ARM_PART_NEOVERSE_V3:
+                return 1;
+        }
+    } else if (implementer == ARM_IMPLEMENTER_QUALCOMM) {
+        /* Qualcomm Oryon (Snapdragon X Elite/Plus) has fast PMULL */
+        if (part == QUALCOMM_PART_ORYON)
+            return 1;
+    }
+#endif
+    return 0;
+}
```
You lied then because you said i had two options.

nmoinvaz commented on cmake/detect-intrinsics.cmake:
```diff
@@ -56,6 +56,40 @@ macro(check_armv8_compiler_flag)
     set(CMAKE_REQUIRED_FLAGS)
 endmacro()
 
+macro(check_armv8_pmull_eor3_compiler_flag)
+    if(NOT NATIVEFLAG)
+        if(CMAKE_C_COMPILER_ID MATCHES "GNU" OR CMAKE_C_COMPILER_ID MATCHES "Clang")
+            check_c_compiler_flag("-march=armv8.2-a+crc+crypto+sha3" HAVE_MARCH_ARMV8_CRYPTO_SHA3)
+            if(HAVE_MARCH_ARMV8_CRYPTO_SHA3)
+                set(PMULLEOR3FLAG "-march=armv8.2-a+crc+crypto+sha3" CACHE INTERNAL "Compiler option to enable ARMv8 PMULL+EOR3 support")
+            else()
+                check_c_compiler_flag("-march=armv8-a+crc+crypto+sha3" HAVE_MARCH_ARMV8A_CRYPTO_SHA3)
+                if(HAVE_MARCH_ARMV8A_CRYPTO_SHA3)
+                    set(PMULLEOR3FLAG "-march=armv8-a+crc+crypto+sha3" CACHE INTERNAL "Compiler option to enable ARMv8 PMULL+EOR3 support")
+                endif()
+            endif()
+        endif()
+    endif()
+    # Check whether compiler supports ARMv8 PMULL + EOR3 intrinsics
+    set(CMAKE_REQUIRED_FLAGS "${PMULLEOR3FLAG} ${NATIVEFLAG} ${ZNOLTOFLAG}")
+    check_c_source_compiles(
+        "#if defined(_MSC_VER) && (defined(_M_ARM64) || defined(_M_ARM64EC))
+        #  include <arm64_neon.h>
+        #else
+        #  include <arm_neon.h>
+        #endif
+        poly128_t f(poly64_t a, poly64_t b) {
+            return vmull_p64(a, b);
+        }
+        uint64x2_t g(uint64x2_t a, uint64x2_t b, uint64x2_t c) {
+            return veor3q_u64(a, b, c);
+        }
+        int main(void) { return 0; }"
+        HAVE_ARMV8_PMULL_EOR3_INTRIN
+    )
+    set(CMAKE_REQUIRED_FLAGS)
+endmacro()
```
It is not necessary because we already have a separate feature check for crc32 for the code. And pmull code doesn't get run unless this feature check passes.

nmoinvaz commented:
Copied the head and tail from `crc32_armv8` for faster `len < 32` processing.

Dead2 commented:
@nmoinvaz This'll need a little rebase

Also, I can confirm that this PR no longer makes things slower on RPI5. 👍

Dead2 commented:
I think this PR also deserves an addition to the Readme.md about support for pmull+eor.

nmoinvaz committed:
Integrate ARMv8 PMULL+EOR3 crc32 algorithm from Peter Cawley

https://github.com/corsix/fast-crc32
https://github.com/zlib-ng/zlib-ng/pull/2023#discussion_r2573303259

Co-authored-by: Peter Cawley <corsix@corsix.org>

nmoinvaz commented:
Rebased and readme updated

Dead2 commented:
> 🧹 Nitpick comments (1)
> > arch/arm/crc32_armv8_pmull_eor3.c (1)> `68-92`: **Consider replacing pointer casts with memcpy to avoid strict-aliasing violations.**
> > > The code casts `uint8_t*` to wider pointer types (`uint16_t*`, `uint32_t*`, `uint64_t*`) and dereferences them. While alignment is checked, this pattern violates C's strict aliasing rules and is technically undefined behavior. Modern compilers optimize fixed-size `memcpy` to single load instructions, making it both safe and performant.
> > > If the project intentionally compiles with `-fno-strict-aliasing`, this can be downgraded. However, using `memcpy` makes the code portable and eliminates reliance on compiler flags.
> > > 
> > > Example safe pattern using memcpy
> > > ```diff
> > > +    uint16_t u16;
> > > +    uint32_t u32;
> > > +    uint64_t u64;
> > > +
> > >      if ((len >= sizeof(uint16_t)) && ((ptrdiff_t)buf & (sizeof(uint32_t) - 1))) {
> > > -        crc0 = __crc32h(crc0, *((uint16_t*)buf));
> > > +        memcpy(&u16, buf, sizeof(u16));
> > > +        crc0 = __crc32h(crc0, u16);
> > >          buf += sizeof(uint16_t);
> > >          len -= sizeof(uint16_t);
> > >      }
> > > 
> > >      if ((len >= sizeof(uint32_t)) && ((ptrdiff_t)buf & (sizeof(uint64_t) - 1))) {
> > > -        crc0 = __crc32w(crc0, *((uint32_t*)buf));
> > > +        memcpy(&u32, buf, sizeof(u32));
> > > +        crc0 = __crc32w(crc0, u32);
> > >          len -= sizeof(uint32_t);
> > >          buf += sizeof(uint32_t);
> > >      }
> > > 
> > >      if (len >= sizeof(uint64_t) && ((ptrdiff_t)buf & (sizeof(uint64_t)))) {
> > > -        crc0 = __crc32d(crc0, *((uint64_t*)buf));
> > > +        memcpy(&u64, buf, sizeof(u64));
> > > +        crc0 = __crc32d(crc0, u64);
> > >          buf += sizeof(uint64_t);
> > >          len -= sizeof(uint64_t);
> > >      }
> > > ```
> > > 
> > > **Note:** Similar casts appear at lines 147-152, 188-193, 217-219, 232, 239, 245, and 250.
> > > Based on learnings, zlib-ng bounds checking is caller-managed, but strict aliasing is a separate portability concern.

Hmm, maybe this comment from coderabbitai is worth fixing @nmoinvaz? IIRC we already have macros that could be used to load the data into a temp var using memcpy.

Some ARM platforms out there do seem to use strict aliasing, so I am not 100% sure this is a problem, but if it can be avoided without downsides then it'd seem like a good idea.

nmoinvaz commented:
I saw that coderabbit comment also, but I decided not to fix it, because it is the exact same code already used in crc32_armv8.c. If I changed it, I would have to change that implementation too and I didn't want to complicate the PR.

nmoinvaz commented on arch/arm/arm_features.c:
```diff
@@ -71,6 +71,113 @@ static int arm_has_crc32(void) {
+#elif defined(_WIN32)
+    /* Windows: No direct API for SHA3, return 0 for now */
+    return 0;
+#elif defined(__ARM_FEATURE_SHA3)
```
Good question. Can you please open a PR?

Andarwinux commented on arch/arm/arm_features.c:
```diff
@@ -71,6 +71,113 @@ static int arm_has_crc32(void) {
+#elif defined(_WIN32)
+    /* Windows: No direct API for SHA3, return 0 for now */
+    return 0;
+#elif defined(__ARM_FEATURE_SHA3)
```
https://github.com/zlib-ng/zlib-ng/pull/2065

brad0 commented:
@nmoinvaz  Looking at this there doesn't seem to be a HWCAP flag of HWCAP2_SHA3. I think you really meant HWCAP2_SVESHA3.

nmoinvaz commented:
I developed this on macOS. So you are probably right. 🥹
